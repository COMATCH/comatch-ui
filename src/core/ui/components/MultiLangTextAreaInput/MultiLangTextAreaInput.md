
## MultiLangTextAreaInput

From [`MultiLangTextAreaInput.js`](MultiLangTextAreaInput.js)



prop | type | default | required | description
---- | :----: | :-------: | :--------: | -----------
**activeLanguage** | `Enum()` |  | :white_check_mark: | 
**name** | `String` |  | :white_check_mark: | 
**onChange** | `Function` |  | :white_check_mark: | 
**value** | `Array[]<Shape>` | `LANGUAGE_CODES.map((lang) => ({ lang, value: '' }))` | :x: | 
**value[].lang** | `String` |  | :x: | 
**value[].value** | `String` |  | :x: | 



