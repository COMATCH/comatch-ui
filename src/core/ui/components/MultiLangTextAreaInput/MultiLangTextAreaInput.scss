.MultiLangTextAreaInput {
    // Hide inactive languages using visibility,
    // this way their values are still submitted
    // as their display is not set to none.
    &.active-lang-en {
        .TextAreaInput:not(.lang-en) {
            visibility: hidden;
            position: absolute;
            z-index: -1;
        }
    }
    &.active-lang-de {
        .TextAreaInput:not(.lang-de) {
            visibility: hidden;
            position: absolute;
            z-index: -1;
        }
    }
    &.active-lang-fr {
        .TextAreaInput:not(.lang-fr) {
            visibility: hidden;
            position: absolute;
            z-index: -1;
        }
    }

    // Resize icons as flags are used
    .TextAreaInput .TextAreaInput__icon,
    .TextAreaInput .InputLabel ~ .TextAreaInput__icon {
        width: 20px;
        height: 16px;
    }
}
