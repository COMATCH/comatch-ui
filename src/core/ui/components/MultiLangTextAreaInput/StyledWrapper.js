import styled from 'styled-components';

export const StyledWrapper = styled.div`
    &.active-lang-en {
        .TextAreaInput:not(.lang-en) {
            visibility: hidden;
            position: absolute;
            z-index: -1;
        }
    }
    &.active-lang-de {
        .TextAreaInput:not(.lang-de) {
            visibility: hidden;
            position: absolute;
            z-index: -1;
        }
    }
    &.active-lang-fr {
        .TextAreaInput:not(.lang-fr) {
            visibility: hidden;
            position: absolute;
            z-index: -1;
        }
    }

    // Resize icons as flags are used
    .TextAreaInput .TextAreaInput__icon,
    .TextAreaInput .InputLabel ~ .TextAreaInput__icon {
        width: 20px;
        height: 16px;
    }
`;
