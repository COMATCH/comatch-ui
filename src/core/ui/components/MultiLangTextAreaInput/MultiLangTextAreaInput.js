import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { languageCodesArray as LANGUAGE_CODES, languageCodesMap as LANGUAGES } from '../../../config';
import { TextAreaInput } from '../TextAreaInput';
import { flagDeBase64 } from './images/flagDeBase64';
import { flagUkBase64 } from './images/flagUkBase64';
import { flagFrBase64 } from './images/flagFrBase64';

import { StyledWrapper } from './StyledWrapper';

export class MultiLangTextAreaInput extends Component {
    static propTypes = {
        activeLanguage: PropTypes.oneOf(LANGUAGE_CODES).isRequired,
        name: PropTypes.string.isRequired,
        onChange: PropTypes.func.isRequired,
        value: PropTypes.arrayOf(
            PropTypes.shape({
                lang: PropTypes.string,
                value: PropTypes.string,
            }),
        ),
    };

    static defaultProps = {
        value: LANGUAGE_CODES.map((lang) => ({ lang, value: '' })),
    };

    renderFlagIcon(lang) {
        const classes = 'TextAreaInput__icon';
        const enFlag = <img src={flagUkBase64} alt="English flag" className={classes} />;
        const deFlag = <img src={flagDeBase64} alt="German flag" className={classes} />;
        const frFlag = <img src={flagFrBase64} alt="French flag" className={classes} />;
        switch (lang) {
            case LANGUAGES.GERMAN:
                return deFlag;
            case LANGUAGES.FRENCH:
                return frFlag;
            case LANGUAGES.ENGLISH:
            default:
                return enFlag;
        }
    }

    getLanguage(name) {
        if (/\.en\./.test(name)) {
            return LANGUAGES.ENGLISH;
        }
        if (/\.de\./.test(name)) {
            return LANGUAGES.GERMAN;
        }
        if (/\.fr\./.test(name)) {
            return LANGUAGES.FRENCH;
        }
        return LANGUAGES.ENGLISH;
    }

    handleTextAreaInputChange = (event) => {
        const { name, value } = event.target;
        const { name: multiLangInputName } = this.props;

        const languageOfInputBeingEdited = this.getLanguage(name);
        const otherValues = this.props.value.filter(({ lang }) => lang !== languageOfInputBeingEdited);
        const multiLangInputValue = [...otherValues, { lang: languageOfInputBeingEdited, value }];

        this.props.onChange({
            ...event,
            target: {
                ...event.target,
                name: multiLangInputName,
                value: multiLangInputValue,
            },
        });
    };

    render() {
        const { activeLanguage, name, onChange, value, ...props } = this.props;
        const classes = classNames('Input', 'MultiLangTextAreaInput', `active-lang-${activeLanguage}`);
        return (
            <StyledWrapper className={classes}>
                {LANGUAGE_CODES.map((language, key) => {
                    const { value: multiLangInputValue } = this.props;
                    const inputData = multiLangInputValue.find(({ lang }) => lang === language);
                    const inputValue = inputData ? inputData.value : '';
                    return (
                        <TextAreaInput
                            className={`lang-${language}`}
                            icon={this.renderFlagIcon(language)}
                            onChange={this.handleTextAreaInputChange}
                            name={`${name}.${language}.value`}
                            value={inputValue}
                            key={key}
                            {...props}
                        />
                    );
                })}
            </StyledWrapper>
        );
    }
}
