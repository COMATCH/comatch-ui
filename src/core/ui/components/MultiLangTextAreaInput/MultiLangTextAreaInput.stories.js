/* eslint-disable import/no-unresolved */
import React, { Component } from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { MultiLangTextAreaInput } from './MultiLangTextAreaInput';

/* eslint-disable react/prop-types */
class Wrapper extends Component {
    state = {
        activeLanguage: 'en',
    };

    setLanguage = (activeLanguage) => this.setState({ activeLanguage });

    render() {
        const { value } = this.props;

        return (
            <div>
                <button type="button" onClick={() => this.setLanguage('en')}>
                    English
                </button>
                <button type="button" onClick={() => this.setLanguage('de')}>
                    German
                </button>
                <button type="button" onClick={() => this.setLanguage('fr')}>
                    French
                </button>
                <hr />
                <MultiLangTextAreaInput
                    label="A name in different language"
                    value={value}
                    activeLanguage={this.state.activeLanguage}
                    onChange={(event) => action('changed')(event.target.name, event.target.value)}
                    name="basicProfile.firstName"
                />
            </div>
        );
    }
}

storiesOf('MultiLangTextAreaInput', module).add('with initial values for all en, de, fr', () => (
    <Wrapper
        value={[
            { lang: 'en', value: 'English name' },
            { lang: 'de', value: 'German name' },
            { lang: 'fr', value: 'French name' },
        ]}
    />
)).add('with initial values only for en', () => (
    <Wrapper
        value={[
            { lang: 'en', value: 'English name' },
        ]}
    />
));
