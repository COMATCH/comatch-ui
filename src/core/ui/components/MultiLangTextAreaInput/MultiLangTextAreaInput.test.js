import React from 'react';
import { mount } from 'enzyme';
import renderer from 'react-test-renderer';
import '../../../config/tests/setup';

import { MultiLangTextAreaInput } from './MultiLangTextAreaInput';

const mockFunction = jest.fn();

const props = {
    activeLanguage: 'en',
    name: 'name',
    onChange: mockFunction,
    value: [{ lang: 'en', value: 'value' }],
};

describe('MultiLangTextAreaInput', () => {
    it('should render without crashing', () => {
        const wrapper = renderer.create(<MultiLangTextAreaInput />).toJSON();
        expect(wrapper).toMatchSnapshot();
    });

    it('should render with default props', () => {
        const wrapper = mount(<MultiLangTextAreaInput />);
        expect(wrapper.prop('value')).toEqual([
            { lang: 'en', value: '' },
            { lang: 'de', value: '' },
            { lang: 'fr', value: '' },
        ]);
    });

    it('should render with custom props', () => {
        const wrapper = mount(<MultiLangTextAreaInput {...props} />);

        expect(wrapper.prop('activeLanguage')).toEqual('en');
        expect(wrapper.prop('name')).toEqual('name');
        expect(wrapper.prop('onChange')).toEqual(mockFunction);
        expect(wrapper.prop('value')).toEqual([{ lang: 'en', value: 'value' }]);
    });
});
