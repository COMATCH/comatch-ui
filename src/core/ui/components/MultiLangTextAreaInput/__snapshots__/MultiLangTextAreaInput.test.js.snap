// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiLangTextAreaInput should render without crashing 1`] = `
<div
  className="Input MultiLangTextAreaInput active-lang-undefined sc-bdVaJa WoCCW"
>
  <div
    className="Input TextAreaInput lang-en block with-icon"
    id={null}
  >
    <textarea
      className="TextAreaInput__input"
      maxLength={null}
      name="undefined.en.value"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      rows={5}
      value=""
    />
    <img
      alt="English flag"
      className="TextAreaInput__icon"
      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTQiIGhlaWdodD0iNDYiIHZpZXdCb3g9IjAgMCA1NCA0NiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB4bWxuczp4bGluaz0iaHR0cDovL3d3dy53My5vcmcvMTk5OS94bGluayI+PGRlZnM+PHBhdGggZD0iTTI1IDIwaDI1djIwTDI1IDIwem0wIDB2MjBIMGwyNS0yMHptMCAwSDBWMGwyNSAyMHptMCAwVjBoMjVMMjUgMjB6IiBpZD0iYSIvPjwvZGVmcz48ZyBmaWxsPSJub25lIiBmaWxsLXJ1bGU9ImV2ZW5vZGQiPjxwYXRoIGZpbGw9IiMwMDI0N0QiIGZpbGwtcnVsZT0ibm9uemVybyIgZD0iTTIgM3Y0MGg1MFYzeiIvPjxwYXRoIGQ9Ik0yIDNsNTAgNDBtMC00MEwyIDQzIiBzdHJva2U9IiNGRkYiIHN0cm9rZS13aWR0aD0iNiIgZmlsbD0iIzAwMCIgZmlsbC1ydWxlPSJub256ZXJvIi8+PGcgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoMiAzKSI+PG1hc2sgaWQ9ImIiIGZpbGw9IiNmZmYiPjx1c2UgeGxpbms6aHJlZj0iI2EiLz48L21hc2s+PHBhdGggZD0iTTAgMGw1MCA0MG0wLTQwTDAgNDAiIHN0cm9rZT0iI0NGMTQyQiIgc3Ryb2tlLXdpZHRoPSI0IiBmaWxsPSIjMDAwIiBmaWxsLXJ1bGU9Im5vbnplcm8iIG1hc2s9InVybCgjYikiLz48L2c+PHBhdGggZD0iTTI3IDN2NDBNMiAyM2g1MCIgc3Ryb2tlPSIjRkZGIiBzdHJva2Utd2lkdGg9IjEwIiBmaWxsPSIjMDAwIiBmaWxsLXJ1bGU9Im5vbnplcm8iLz48cGF0aCBkPSJNMjcgM3Y0ME0yIDIzaDUwIiBzdHJva2U9IiNDRjE0MkIiIHN0cm9rZS13aWR0aD0iNiIgZmlsbD0iIzAwMCIgZmlsbC1ydWxlPSJub256ZXJvIi8+PC9nPjwvc3ZnPg=="
    />
  </div>
  <div
    className="Input TextAreaInput lang-de block with-icon"
    id={null}
  >
    <textarea
      className="TextAreaInput__input"
      maxLength={null}
      name="undefined.de.value"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      rows={5}
      value=""
    />
    <img
      alt="German flag"
      className="TextAreaInput__icon"
      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAiIGhlaWdodD0iNDAiIHZpZXdCb3g9IjAgMCA1MCA0MCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsLXJ1bGU9Im5vbnplcm8iIGZpbGw9Im5vbmUiPjxwYXRoIGZpbGw9IiMwMDAiIGQ9Ik0wIDBoNTB2NDBIMHoiLz48cGF0aCBmaWxsPSIjRDAwIiBkPSJNMCAxMy4zMzNoNTBWNDBIMHoiLz48cGF0aCBmaWxsPSIjRkZDRTAwIiBkPSJNMCAyNi42NjdoNTBWNDBIMHoiLz48L2c+PC9zdmc+"
    />
  </div>
  <div
    className="Input TextAreaInput lang-fr block with-icon"
    id={null}
  >
    <textarea
      className="TextAreaInput__input"
      maxLength={null}
      name="undefined.fr.value"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      placeholder=""
      rows={5}
      value=""
    />
    <img
      alt="French flag"
      className="TextAreaInput__icon"
      src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iNTAiIGhlaWdodD0iNDAiIHZpZXdCb3g9IjAgMCA1MCA0MCIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj48ZyBmaWxsLXJ1bGU9Im5vbnplcm8iIGZpbGw9Im5vbmUiPjxwYXRoIGZpbGw9IiNFRDI5MzkiIGQ9Ik0wIDBoNTB2NDBIMHoiLz48cGF0aCBmaWxsPSIjRkZGIiBkPSJNMCAwaDMzLjMzM3Y0MEgweiIvPjxwYXRoIGZpbGw9IiMwMDIzOTUiIGQ9Ik0wIDBoMTYuNjY3djQwSDB6Ii8+PC9nPjwvc3ZnPg=="
    />
  </div>
</div>
`;
